File: scripts/skill-paths.mjs
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const projectRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..')

/** Return the finite directory ancestry used to discover workspace skills. */
const ancestorDirectoriesOf = (directory) => {
  const parent = path.dirname(directory)
  return parent === directory
    ? [directory]
    : [directory, ...ancestorDirectoriesOf(parent)]
}

const configuredRoots = [
  process.env.AGENT_SKILLS_ROOT,
  process.env.CODEX_SKILLS_ROOT,
].filter(Boolean)

const discoveredRoots = [
  ...ancestorDirectoriesOf(projectRoot).map((root) =>
    path.join(root, '.agents', 'skills')
  ),
  path.join(os.homedir(), '.agents', 'skills'),
  path.join(os.homedir(), '.codex', 'skills'),
]

const skillRoots = [...new Set([...configuredRoots, ...discoveredRoots])]

/** Resolve an installed skill directory without embedding one developer home. */
export const skillDirectoryOf = (directoryName) => {
  const candidates = skillRoots.map((root) => path.join(root, directoryName))
  return candidates.find((candidate) => fs.existsSync(candidate)) ?? candidates[0]
}

/** Resolve an installed skill entrypoint without embedding one checkout path. */
export const skillFileOf = (directoryName) =>
  path.join(skillDirectoryOf(directoryName), 'SKILL.md')

File: scripts/fp/skill_paths.py
"""Portable discovery for the installed FP skill authority."""

from __future__ import annotations

import os
from pathlib import Path


PROJECT_ROOT = Path(__file__).resolve().parents[2]


def _configured_roots() -> tuple[Path, ...]:
    values = (
        os.environ.get("AGENT_SKILLS_ROOT"),
        os.environ.get("CODEX_SKILLS_ROOT"),
    )
    return tuple(Path(value).expanduser() for value in values if value)


def _discovered_roots() -> tuple[Path, ...]:
    workspace = tuple(parent / ".agents" / "skills" for parent in (
        PROJECT_ROOT,
        *PROJECT_ROOT.parents,
    ))
    home = (
        Path.home() / ".agents" / "skills",
        Path.home() / ".codex" / "skills",
    )
    return workspace + home


def skill_file(directory_name: str) -> Path:
    """Resolve one installed skill entrypoint from configured and known roots."""
    roots = tuple(dict.fromkeys(_configured_roots() + _discovered_roots()))
    candidates = tuple(root / directory_name / "SKILL.md" for root in roots)
    return next((candidate for candidate in candidates if candidate.is_file()), candidates[0])
